keps/sig-instrumentation/20190605-metrics-validation-and-verification.md

```diff
@@ -0,0 +1,180 @@
+---
+title: Metrics Validation and Verification
+authors:
+  - "@serathius"
+  - "@solodov"
+  - "@logicalhan"
+owning-sig: sig-instrumentation
+participating-sigs:
+  - sig-instrumentation
+  - sig-testing
+reviewers:
+  - "@brancz"
+approvers:
+  - "@brancz"
+editor: TBD
+creation-date: 2019-06-05
+last-updated: 2019-06-05
+status: implementable 
+see-also:
+  - 20190404-kubernetes-control-plane-metrics-stability
+  - 20190605-metrics-stability-migration
+---
+
+ # Metrics Validation and Verification
+
+ ## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+- [Design Details](#design-details)
+  - [Static Analysis](#static-analysis)
+  - [Discovery of golang source files](#discovery-of-golang-source-files)
+  - [Format of defining stable metrics](#format-of-defining-stable-metrics)
+  - [Stable metric detection strategy](#stable-metric-detection-strategy)
+  - [Format of stable metrics list file](#format-of-stable-metrics-list-file)
+  - [Failure Modes](#failure-modes)
+  - [Performance](#performance)
+- [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Test Plan](#test-plan)
+- [Unresolved questions](#unresolved-questions)
+  - [Should changes in histogram buckets/summary objectives be a breaking change?](#should-changes-in-histogram-bucketssummary-objectives-be-a-breaking-change)
+  - [Should code snippets used for unit testing be kept in separate files instead of strings?](#should-code-snippets-used-for-unit-testing-be-kept-in-separate-files-instead-of-strings)
+<!-- /toc -->
+
+## Summary
+
+ This Kubernetes Enhancement Proposal (KEP) builds off of the framework proposed
+in the [Kubernetes Control-Plane Metrics Stability
+KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/20190404-kubernetes-control-plane-metrics-stability.md)
+and proposes a strategy for ensuring conformance of metrics with official
+stability guarantees.
+
+## Motivation
+
+ While the [Kubernetes Control Plane metrics stability KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/20190404-kubernetes-control-plane-metrics-stability.md) provides a framework to define stability levels for control-plane metrics, it does not provide a strategy for verifying and validating conformance to stated guarantees. This KEP intends to propose a framework for validating and verifying metric guarantees.
+
+### Goals
+
+* Given a stable metric, validate that we cannot remove or modify it (other than adding deprecation information).
+* Given a deprecated but stable metric, validate that we cannot remove or modify it until the deprecation period has elapsed.
```

Comment:
Where is this deprecation period stated, I haven't seen it mentioned in this doc (could have missed it). I have seen in the original KEP, as `DeprecatedVersion`, is this what this is referring to. If yes, any plans on how will we check in which kubernetes release this metric will land to make sure we can remove it.

Reply:
It might be worth reiterating in this document, but if I recall correctly the version a metric was first deprecated in is part of the metric's metadata.

Reply:
Yes, DepracatedVersion is part of metadata put in `stable-metrics-list.yaml`. This means that decision of deprecating metric will require approval from sig-instrumentation.